lib/cfp.js

```diff
@@ -0,0 +1,325 @@
+  const savedHashtag = await collection.where('hashtag', '==', hashtag).get();
+  if (savedHashtag.empty) {
+    await collection.add({
+      hashtag,
+      tweetId: tweetDocId,
+      count: 1,
+    });
+  } else {
+    const [doc] = savedHashtag.docs;
+    await collection.doc(doc.id).update({
+      count: doc.data().count + 1,
+    });
+  }
+};
+
+const saveTweetHashtags = async (formattedTweet, db = null) => {
+  if (!formattedTweet.doc) {
+    throw new Error('You first need to save the tweet');
+  }
+
+  const { hashtags, doc: { id: tweetDocId } } = formattedTweet;
+  await Promise.all(hashtags.map(async (hashtag) => {
+    await saveHashtag(hashtag, tweetDocId, db);
+  }));
+};
+
+const saveTweetsHashtags = async (formattedTweets, db = null) => formattedTweets.reduce(
+  (currentPromise, tweet) => currentPromise.then(() => saveTweetHashtags(
+    tweet,
+    db,
+  )),
+  Promise.resolve(),
+);
+
+const saveHashtagCount = async (hashtag, count, db = null) => {
+  const useDB = db || getDB();
+  const collection = useDB.collection(COLLECTIONS.COUNTERS);
+  const savedCounter = await collection.where('hashtag', '==', hashtag).get();
+  if (savedCounter.empty) {
+    await collection.add({
+      hashtag,
+      count,
+    });
+  } else {
+    const [doc] = savedCounter.docs;
+    await collection.doc(doc.id).update({ count });
+  }
+
+  return {
+    hashtag,
+    count,
+  };
+};
+
+const processHashtags = async (db = null) => {
+  const useDB = db || getDB();
+  const hashtagsResult = await useDB.collection(COLLECTIONS.HASHTAGS).get();
+  const hashtags = hashtagsResult.docs.reduce(
+    (acc, doc) => {
+      const data = doc.data();
+      return {
+        ...acc,
+        [data.hashtag]: data.count,
+      };
+    },
+    {},
+  );
+  const entriesResult = await useDB.collection(COLLECTIONS.ENTRIES).get();
+  let calculated = entriesResult.docs.reduce(
+    (acc, doc) => {
+      const data = doc.data();
+      const nextAcc = { ...acc };
+      if (nextAcc[data.hashtag]) {
+        nextAcc[data.hashtag] += data.value;
+      } else {
+        nextAcc[data.hashtag] = data.value;
+      }
+
+      return nextAcc;
+    },
+    hashtags,
+  );
+  const aliasesResult = await useDB.collection(COLLECTIONS.ALIASES).get();
+  calculated = aliasesResult.docs.reduce(
+    (acc, doc) => {
+      let nextAcc;
+      const data = doc.data();
+      if (acc[data.from]) {
+        nextAcc = { ...acc };
+        if (nextAcc[data.hashtag]) {
+          nextAcc[data.hashtag] += nextAcc[data.from];
+        } else {
+          nextAcc[data.hashtag] = nextAcc[data.from];
+        }
+
+        delete nextAcc[data.from];
+      } else {
+        nextAcc = acc;
+      }
+
+      return nextAcc;
+    },
+    calculated,
+  );
+
+  const results = await Promise.all(Object.keys(calculated).map(async (key) => {
+    const newCount = await saveHashtagCount(key, calculated[key], useDB);
+    return newCount;
+  }));
```

Comment:
You could get rid of the `newCount` variable, and the inner `await` is unnecessary while wrapping on a `Promise.all()`:

```js
const results = await Promise.all(
  Object.keys(calculated).map(
    async (key) => saveHashtagCount(key, calculated[key], useDB)
  )
);
```